File: examples/lg/pmbgp.c
/*  
    pmacct (Promiscuous mode IP Accounting package)
    pmacct is Copyright (C) 2003-2020 by Paolo Lucente
*/

/*
    This program is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation; either version 2 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program; if not, write to the Free Software
    Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.
*/

#define __PMBGP_C

/* includes */
#include "pmacct.h"
#include "pmacct-data.h"
#include "addr.h"
#ifdef WITH_ZMQ
#include "zmq_common.h"
#endif
#include "bgp/bgp.h"
#include "pmbgpd.h"
#include "pmbgp.h"

/* functions */
#if defined (WITH_ZMQ) && defined (WITH_JANSSON) 
void usage_pmbgp(char *prog)
{
  printf("%s %s (%s)\n", PMBGP_USAGE_HEADER, PMACCT_VERSION, PMACCT_BUILD);
  printf("Usage: %s [options] [query]\n\n", prog);
  printf("IP Lookup query options:\n");
  printf("  -a\tIP address/prefix to look up\n");
  printf("  -d\tRoute Distinguisher to look up\n");
  printf("  -r\tBGP peer to look up\n");
  printf("  -R\tTCP port of the BGP peer (for BGP through NAT/proxy scenarios)\n");
  printf("\n");
  printf("Get Peers query options:\n");
  printf("  -g\tGet the list of BGP peers at the Looking Glass\n");
  printf("\n");
  printf("General options:\n");
  printf("  -z\tLooking Glass IP address [default: 127.0.0.1]\n");
  printf("  -Z\tLooking Glass port [default: 17900]\n");
  printf("  -u\tLooking Glass username [default: none]\n");
  printf("  -p\tLooking Glass password [default: none]\n");
  printf("\n");
  printf("  -h\tShow this page\n");
  printf("  -V\tPrint version and exit\n");
}

void version_pmbgp(char *prog)
{
  printf("%s %s (%s)\n", PMBGP_USAGE_HEADER, PMACCT_VERSION, PMACCT_BUILD);
}

int main(int argc,char **argv)
{
  char address_str[SRVBUFLEN], peer_str[SRVBUFLEN], rd_str[SRVBUFLEN], port_str[SRVBUFLEN];
  char *req_str = NULL, *req_type_str = NULL, *rep_str = NULL, *pfx_delim = NULL;
  char *zmq_host_str_ptr, zmq_host_str[SRVBUFLEN], default_zmq_host_str[] = "127.0.0.1";
  int zmq_port = 0, default_zmq_port = 17900, results = 0, query_type = 0, idx = 0;

  struct p_zmq_host zmq_host;
  struct host_addr peer_ha, address_ha;

  /* getopt() stuff */
  extern char *optarg;
  extern int optind, opterr, optopt;
  int cp, ip_lookup_query, get_peers_query;

  memset(address_str, 0, sizeof(address_str));
  memset(rd_str, 0, sizeof(rd_str));
  memset(peer_str, 0, sizeof(peer_str));
  memset(port_str, 0, sizeof(port_str));
  memset(zmq_host_str, 0, sizeof(zmq_host_str));
  memset(&zmq_host, 0, sizeof(zmq_host));
  memset(&address_ha, 0, sizeof(address_ha));
  memset(&peer_ha, 0, sizeof(peer_ha));
  ip_lookup_query = FALSE;
  get_peers_query = FALSE;

  while ((cp = getopt(argc, argv, ARGS_PMBGP)) != -1) {
    switch (cp) {
    case 'h':
      usage_pmbgp(argv[0]);
      exit(0);
      break;
    case 'V':
      version_pmbgp(argv[0]);
      exit(0);
      break;
    case 'a':
      strlcpy(address_str, optarg, sizeof(address_str));
      ip_lookup_query = TRUE;
      break;
    case 'd':
      strlcpy(rd_str, optarg, sizeof(rd_str));
      ip_lookup_query = TRUE;
      break;
    case 'r':
      strlcpy(peer_str, optarg, sizeof(peer_str));
      ip_lookup_query = TRUE;
      break;
    case 'R':
      strlcpy(port_str, optarg, sizeof(port_str));
      ip_lookup_query = TRUE;
      break;
    case 'g':
      get_peers_query = TRUE;
      break;
    case 'z':
      strlcpy(zmq_host_str, optarg, sizeof(zmq_host_str));
      break;
    case 'Z':
      zmq_port = atoi(optarg);
      break;
    case 'u':
      strlcpy(zmq_host.zap.username, optarg, sizeof(zmq_host.zap.username));
      break;
    case 'p':
      strlcpy(zmq_host.zap.password, optarg, sizeof(zmq_host.zap.password));
      break;
    default:
      printf("ERROR: parameter %c unknown! \n  Exiting...\n\n", cp);
      usage_pmbgp(argv[0]);
      exit(1);
      break;
    }
  }

  if (!ip_lookup_query && !get_peers_query) {
    printf("ERROR: no query specificed. Exiting ..\n");
    usage_pmbgp(argv[0]);
    exit(1);
  }

  if (ip_lookup_query && get_peers_query) {
    printf("ERROR: IP Lookup and Get Peers queries are mutual exclusive. Please select only one. Exiting ..\n");
    exit(1);
  }

  if (ip_lookup_query && (!strlen(address_str) || !strlen(peer_str))) {
    printf("ERROR: mandatory options for IP Lookup query (-a,  -r) are not specified. Exiting ..\n");
    exit(1);
  }
  
  if (!strlen(zmq_host_str)) zmq_host_str_ptr = default_zmq_host_str; 
  else zmq_host_str_ptr = zmq_host_str; 

  if (!zmq_port) zmq_port = default_zmq_port;

  /* craft query */
  if (ip_lookup_query) {
    {
      json_t *req_obj = json_object();

      json_object_set_new_nocheck(req_obj, "query_type", json_integer(BGP_LG_QT_IP_LOOKUP));
      json_object_set_new_nocheck(req_obj, "queries", json_integer(1));
      req_type_str = json_dumps(req_obj, JSON_PRESERVE_ORDER);
      json_decref(req_obj);
    }

    {
      json_t *req_obj = json_object();

      str_to_addr(peer_str, &peer_ha);
      if (peer_ha.family) json_object_set_new_nocheck(req_obj, "peer_ip_src", json_string(peer_str));
      else {
        printf("ERROR: invalid -r value. Exiting ..\n");
        exit(1);
      }

      if (strlen(port_str)) {
	int port_int;

	port_int = atoi(port_str);
	if (port_int > 0 && port_int <= 65535) json_object_set_new_nocheck(req_obj, "peer_tcp_port", json_integer(port_int));
	else {
	  printf("ERROR: invalid -R value. Exiting ..\n");
	  exit(1);
	}
      }

      /* Simplified validation: only IP address part */
      pfx_delim = strchr(address_str, '/');
      if (pfx_delim) (*pfx_delim) = '\0'; 
      str_to_addr(address_str, &address_ha);
      if (pfx_delim) (*pfx_delim) = '/'; 

      if (address_ha.family) json_object_set_new_nocheck(req_obj, "ip_prefix", json_string(address_str));
      else {
	printf("ERROR: invalid -a value. Exiting ..\n");
	exit(1);
      }

      /* no specific validation done for the RD */
      if (strlen(rd_str)) json_object_set_new_nocheck(req_obj, "rd", json_string(rd_str));

      req_str = json_dumps(req_obj, JSON_PRESERVE_ORDER);
      json_decref(req_obj);
    }
  }
  else if (get_peers_query) {
    {
      json_t *req_obj = json_object();

      json_object_set_new_nocheck(req_obj, "query_type", json_integer(BGP_LG_QT_GET_PEERS));
      json_object_set_new_nocheck(req_obj, "queries", json_integer(1));
      req_type_str = json_dumps(req_obj, JSON_PRESERVE_ORDER);
      json_decref(req_obj);
    }
  }

  pmbgp_zmq_req_setup(&zmq_host, zmq_host_str_ptr, zmq_port);

  if (req_str) {
    pmbgp_zmq_sendmore_str(&zmq_host.sock, req_type_str);
    pmbgp_zmq_send_str(&zmq_host.sock, req_str);
  }
  else {
    pmbgp_zmq_send_str(&zmq_host.sock, req_type_str);
  }

  /* query type + results */
  rep_str = pmbgp_zmq_recv_str(&zmq_host.sock);
  if (rep_str) {
    json_error_t rep_err;
    json_t *rep_results_obj, *results_json, *query_type_json;

    rep_results_obj = json_loads(rep_str, 0, &rep_err);

    if (rep_results_obj) {
      if (!json_is_object(rep_results_obj)) {
        printf("WARN: json_is_object() failed for results: %s\n", rep_err.text);
	exit(1);
      }
      else {
        query_type_json = json_object_get(rep_results_obj, "query_type");
        if (query_type_json == NULL) {
          printf("WARN: no 'query_type' element.\n");
          exit(1);
        }
        else query_type = json_integer_value(query_type_json);

        results_json = json_object_get(rep_results_obj, "results");
        if (results_json == NULL) {
          printf("WARN: no 'results' element.\n");
	  exit(1);
	}
        else results = json_integer_value(results_json);
      }

      json_decref(rep_results_obj);
    }

    printf("%s\n", rep_str);
    free(rep_str);
  }
  
  /* data */
  for (idx = 0; idx < results; idx++) {
    rep_str = pmbgp_zmq_recv_str(&zmq_host.sock);
    if (rep_str) {
      if (query_type == BGP_LG_QT_IP_LOOKUP ||
	  query_type == BGP_LG_QT_GET_PEERS)
	printf("%s\n", rep_str);
      free(rep_str);
    }
  }

  return 0;
}

void pmbgp_zmq_req_setup(struct p_zmq_host *zmq_host, char *host, int port)
{
  int ret;

  if (!zmq_host->ctx) zmq_host->ctx = zmq_ctx_new();

  zmq_host->sock.obj = zmq_socket(zmq_host->ctx, ZMQ_REQ);
  if (!zmq_host->sock.obj) {
    printf("ERROR: zmq_socket() failed for ZMQ_REQ: %s. Exiting.\n", zmq_strerror(errno));
    exit(1);
  }

  snprintf(zmq_host->sock.str, sizeof(zmq_host->sock.str), "tcp://%s:%u", host, port);

  if (strlen(zmq_host->zap.username) && strlen(zmq_host->zap.password)) {
    ret = zmq_setsockopt(zmq_host->sock.obj, ZMQ_PLAIN_USERNAME, zmq_host->zap.username, strlen(zmq_host->zap.username));
    if (ret == ERR) {
      printf("ERROR: zmq_setsockopt() ZMQ_PLAIN_USERNAME failed: %s\nExiting.\n", zmq_strerror(errno));
      exit(1);
    }

    ret = zmq_setsockopt(zmq_host->sock.obj, ZMQ_PLAIN_PASSWORD, zmq_host->zap.password, strlen(zmq_host->zap.password));
    if (ret == ERR) {
      printf("ERROR: zmq_setsockopt() ZMQ_PLAIN_PASSWORD failed: %s\nExiting.\n", zmq_strerror(errno));
      exit(1);
    }
  }

  ret = zmq_connect(zmq_host->sock.obj, zmq_host->sock.str);
  if (ret == ERR) {
    printf("ERROR: zmq_connect() failed for ZMQ_REQ: %s. Exiting.\n", zmq_strerror(errno));
    exit(1);
  }
}

char *pmbgp_zmq_recv_str(struct p_zmq_sock *sock)
{
  char buf[LARGEBUFLEN];
  int len;

  memset(buf, 0, sizeof(buf));
  len = zmq_recv(sock->obj, buf, (sizeof(buf) - 1), 0);
  if (len == ERR) return NULL;
  else return strndup(buf, sizeof(buf));
}

int pmbgp_zmq_send_str(struct p_zmq_sock *sock, char *buf)
{
  int len;

  len = zmq_send(sock->obj, buf, strlen(buf), 0);

  return len;
}

int pmbgp_zmq_sendmore_str(struct p_zmq_sock *sock, char *buf)
{
  int len;

  len = zmq_send(sock->obj, buf, strlen(buf), ZMQ_SNDMORE);

  return len;
}
#else
int main(int argc,char **argv)
{
  printf("WARN: pmbgp: tool depends on missing --enable-zmq and --enable-jansson. Exiting.\n");
  return 1;
}
#endif

File: examples/lg/pmbgp.py
#!/usr/bin/env python3
#
#

import sys, os, getopt, struct, time
import json
import zmq

def usage(tool):
    print("")
    print(f"Usage: {tool} [options] [query]")
    print("")

    print("IP Lookup query options:")
    print("  -a, --prefix".ljust(25) + "IP address/prefix to look up")
    print("  -d, --rd".ljust(25) + "Route Distinguisher to look up")
    print("  -r, --peer".ljust(25) + "BGP peer to look up")
    print("  -R, --peer-port".ljust(25) + "TCP port of the BGP peer (for BGP through NAT/proxy scenarios)")
    print("")
    print("Get Peers query options:")
    print("  -g, --get-peers".ljust(25) + "Get the list of BGP peers at the Looking Glass")
    print("")
    print("General options:")
    print("  -z, --zmq-host".ljust(25) + "Looking Glass IP address [default: 127.0.0.1]")
    print("  -Z, --zmq-port".ljust(25) + "Looking Glass port [default: 17900]")
    print("  -u, --zmq-user".ljust(25) + "Looking Glass username [default: none]")
    print("  -p, --zmq-passwd".ljust(25) + "Looking Glass password [default: none]")
    print("")
    print("  -h, --help".ljust(25) + "Print this help")

def main():
    try:
        opts, args = getopt.getopt(sys.argv[1:], "ha:d:r:R:z:Z:u:p:g", ["help", "prefix=", "rd=",
                "peer=", "peer-port=", "zmq-host=", "zmq-port=", "zmq-user=", "zmq-passwd=",
                "get-peers="])
    except getopt.GetoptError as err:
        print(str(err))
        usage(sys.argv[0])
        sys.exit(2)

    # Vars for connecting to the LG
    lgHost = "127.0.0.1"
    lgPort = "17900"
    lgUser = b"" 
    lgPasswd = b"" 
    queryTypeObj = {}
    queryDataObj = {}
    queryTypeJsonObj = ""
    queryDataJsonObj = ""

    # Vars for IP Lookup query
    queryIpl = 0
    queryIplPrefix = "" 
    queryIplRd = "" 
    queryIplPeer = "" 
    queryIplPeerPort = ""
    queryIplRequiredCl = 0

    # Vars for Get Peers query
    queryGp = 0
    
    for o, a in opts:
        if o in ("-h", "--help"):
            usage(sys.argv[0])
            sys.exit()
        elif o in ("-a", "--prefix"):
            queryIplRequiredCl += 1
            queryIpl = 1
            queryIplPrefix = a
        elif o in ("-d", "--rd"):
            queryIpl = 1
            queryIplRd = a
        elif o in ("-r", "--peer"):
            queryIplRequiredCl += 1
            queryIpl = 1
            queryIplPeer = a
        elif o in ("-R", "--peer-port"):
            queryIpl = 1
            queryIplPeerPort = a
        elif o in ("-g", "--get-peers"):
            queryGp = 1
        elif o in ("-z", "--zmq-host"):
            lgHost = a
        elif o in ("-Z", "--zmq-port"):
            lgPort = a
        elif o in ("-u", "--zmq-user"):
            lgUser = a.encode()
        elif o in ("-p", "--zmq-passwd"):
            lgPasswd = a.encode()
        else:
            assert False, "unhandled option"

    # Validations
    if not queryGp and not queryIpl:
        print("ERROR: no query specificed.")
        usage(sys.argv[0])
        sys.exit(1)

    if queryGp and queryIpl:
        print("ERROR: IP Lookup and Get Peers queries are mutual exclusive. Please select only one.")
        sys.exit(1)

    if queryIpl and queryIplRequiredCl < 2: 
        print("ERROR: Missing required arguments (-a, -r) for IP Lookup query")
        usage(sys.argv[0])
        sys.exit(1)

    # Craft query
    if queryIpl:
        queryTypeObj['query_type'] = 1
        queryTypeObj['queries'] = 1
        queryDataObj['peer_ip_src'] = queryIplPeer
        queryDataObj['ip_prefix'] = queryIplPrefix

        if len(queryIplRd):
            queryDataObj['rd'] = queryIplRd

        if len(queryIplPeerPort):
            queryDataObj['peer_tcp_port'] = queryIplPeerPort

        queryTypeJsonObj = json.dumps(queryTypeObj).encode()
        queryDataJsonObj = json.dumps(queryDataObj).encode()


    if queryGp:
        queryTypeObj['query_type'] = 2
        queryTypeObj['queries'] = 1
        queryTypeJsonObj = json.dumps(queryTypeObj).encode()


    # Connect to LG, send request, read reply 
    lgClientZmqCtx = zmq.Context()
    lgClientZmqReq = lgClientZmqCtx.socket(zmq.REQ)

    if len(lgUser):
        lgClientZmqReq.plain_username = lgUser

    if len(lgPasswd):
        lgClientZmqReq.plain_password = lgPasswd

    lgConnectStr = "tcp://" + lgHost + ":" + lgPort
    lgClientZmqReq.connect(lgConnectStr)

    if len(queryDataJsonObj):
        lgClientZmqReq.send(queryTypeJsonObj, zmq.SNDMORE)
        lgClientZmqReq.send(queryDataJsonObj)
    else:
        lgClientZmqReq.send(queryTypeJsonObj)

    repResults = lgClientZmqReq.recv();
    if len(repResults):
        repType = 0
        repNum = 0

        repResultsJsonObj = json.loads(repResults)
        if not 'query_type' in repResultsJsonObj.keys():
            print("WARN: no 'query_type' element.")
            sys.exit(1)
        else:
            repType = repResultsJsonObj['query_type']

        if not 'results' in repResultsJsonObj.keys():
            print("WARN: no 'results' element.")
            sys.exit(1)
        else:
            repNum = repResultsJsonObj['results']

        print(repResults)

        repIdx = 0
        while repIdx < repNum:
            repData = lgClientZmqReq.recv();
            print(repData)

            repIdx += 1

if __name__ == "__main__":
    main()
